### ISSUE TITLE
`reportIncompatibleVariableOverride` for additive overrides

### ISSUE THREAD

**Issue by dlucredativ** [microsoft/pyright#11258]
The following code

```
class BaseOuter:
    class Inner:
        field1: int
    inner: Inner

class Outer(BaseOuter):
    class Inner(BaseOuter.Inner):
        field2: str
    inner: Inner
```
triggers
```
/tmp/test.py
  /tmp/test.py:9:5 - error: "inner" overrides symbol of same name in class "BaseOuter"
    Variable is mutable so its type is invariant
      Override type "Inner" is not the same as base type "Inner" (reportIncompatibleVariableOverride)
1 error, 0 warnings, 0 informations
```

Is this really not type-safe?

**Comment by erictraut** [comment]
Pyright is working correctly here. The class `Outer` is redefining the type of the symbol `Inner` in an incompatible manner. This violates the Liskov Substitution Principle and is therefore not type safe.

**Comment by dlucredativ** [comment]
Could you please elaborate how the LSP is violated? As I currently understand it, `BaseOuter` can only have the following expectations:

```
class BaseOuter:
    class Inner:
        field1: int = 1
    inner: Inner = Inner()

class Outer(BaseOuter):
    class Inner(BaseOuter.Inner):
        field2: str = "2"
    inner: Inner = Inner()

assert issubclass(BaseOuter, BaseOuter)
assert issubclass(Outer, BaseOuter)
assert isinstance(BaseOuter(), BaseOuter)
assert isinstance(Outer(), BaseOuter)

assert issubclass(BaseOuter.Inner, BaseOuter.Inner)
assert issubclass(Outer.Inner, BaseOuter.Inner)
assert isinstance(BaseOuter.Inner(), BaseOuter.Inner)
assert isinstance(Outer.Inner(), BaseOuter.Inner)

assert isinstance(BaseOuter.Inner.field1, int)
assert isinstance(Outer.Inner.field1, int)
```

**Comment by erictraut** [comment]
```python
def func(x: BaseOuter):
    print(x.inner.field1)

o = Outer()
o.inner = Outer.Inner()
func(o)  # Crash
```

ChatGPT can also provide additional explanation or examples if you'd like.

**Comment by dlucredativ** [comment]
This crash has nothing to do with `Outer`:
```
class BaseOuter:
    class Inner:
        field1: int
    inner: Inner

def func(x: BaseOuter) -> None:
    print(x.inner.field1)

bo = BaseOuter()                                                                                                                                                                                                                                                                                      
bo.inner = BaseOuter.Inner()
func(bo)  # Crash
```
No complaints from `pyright` there.

With preset `field1`:

```
class BaseOuter:
    class Inner:
        field1: int = 0
    inner: Inner

class Outer(BaseOuter):
    class Inner(BaseOuter.Inner):
        field2: str = "str"
    inner: Inner


def func(x: BaseOuter) -> None:
    print(x.inner.field1)                                                                                                                                                                                                                                                                             


o = BaseOuter()
o.inner = BaseOuter.Inner()
func(o)  # No Crash
```

**Comment by erictraut** [comment]
```python
class BaseOuter:
    class Inner:
        field1: int = 0

    inner: Inner

    def reset(self):
        self.inner = BaseOuter.Inner()


class Outer(BaseOuter):
    class Inner(BaseOuter.Inner):
        field2: str = "str"

    inner: Inner

    def print_field2(self):
        print(self.inner.field2)


def func(x: BaseOuter) -> None:
    print(x.inner.field1)


o = Outer()

o.print_field2()  # No crash

o.reset()
func(o)
o.print_field2()  # Crash
```

Because the `inner` variable is mutable, its type is invariant, not covariant. If it were marked read-only, then this code would be type safe.

Pyright (correctly) produces the same error even when there is no local `class` definition involved.

```python
class A:
    x: Sequence[int]

class B(A):
    x: list[int]
```

**Comment by dlucredativ** [comment]
Thanks, this helped. So I need to mark `BaseOuter.inner` as readonly somehow.
